// metals-languageclient/lib/interfaces/UserConfiguration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserConfiguration = void 0;
/**
 * These are some of the common user configuration options used in Metals and needed for the editors.
 *
 *  - https://scalameta.org/metals/docs/integrations/new-editor#metals-user-configuration
 */
var UserConfiguration;
(function (UserConfiguration) {
    /**
     * Metals server version
     */
    UserConfiguration["ServerVersion"] = "serverVersion";
    /**
     * Metals server Java properties
     */
    UserConfiguration["ServerProperties"] = "serverProperties";
    /**
     * Directory containing the Java binary.
     */
    UserConfiguration["JavaHome"] = "javaHome";
    /**
     * Additional repositories that can be used to download dependencies.
     * https://get-coursier.io/docs/other-repositories
     */
    UserConfiguration["CustomRepositories"] = "customRepositories";
    /**
     * Repository to use instead of maven central for example `https://jcenter.bintray.com`
     */
    UserConfiguration["CoursierMirror"] = "coursierMirror";
})(UserConfiguration = exports.UserConfiguration || (exports.UserConfiguration = {}));
